### ISSUE TITLE
EventTransformer をモダリティ別 Transformer＋Cross‐Attention プーリング構造に改修

### ISSUE THREAD

**Issue by Motoki0705** [Motoki0705/Tennis-Analyzer#13]
### 背景
現状の EventTransformer では、各フレームにおける ball, court, players 情報を以下の手順でまとめていました：

1. 各フレームの ball_embed, court_embed, masked mean pooling した player_repr をそれぞれ計算  
2. その 3 つを concat → Linear（3d_model → d_model）で射影  
3. LearnablePositionalEncoding → TransformerEncoder → Linear（分類ヘッド）

ただし以下のような情報損失懸念があります：
- **プレイヤー情報を「masked mean」で単純平均**すると、複数のプレイヤーが存在する場合に細かい違いが埋もれてしまう。  
- **ball, court, player を concat→線形投影した後にまとめる**ため、「どのモダリティをどれだけ重視すべきか」を学習できない。  

### 変更内容
以下の 3 段階でアーキテクチャを改修します。

1. **プレイヤー情報に Attention Pooling を導入**  
   - `Masked Mean` の代わりに、`MultiheadAttention` を使って各フレーム内のプレイヤー埋め込みを「重み付き加重平均」します。  
   - クエリは学習可能な埋め込みクエリ（`player_attn_q`）を用い、キー／値には player 埋め込みを与えて「どのプレイヤーを重視するか」を動的に学習します。  

2. **ball/court/player ごとに独立した小さな TransformerEncoder を適用**  
   - 各モダリティ（ball, court, player）が固有に持つ時系列特徴を先に抽出するため、  
     ```python
     x_ball   = encoder_ball(ball_repr)    # (B, T, d_model)
     x_court  = encoder_court(court_repr)  # (B, T, d_model)
     x_player = encoder_player(player_repr) # (B, T, d_model)
     ```  
     のようにそれぞれ独立したトランスフォーマを経由させます。これにより「モダリティ固有の自己相互作用」をまずモデル化できるようになります。

3. **3 モダリティを Cross‐Attention で融合**  
   - 上記で得られた `x_ball`, `x_court`, `x_player` （いずれも (B, T, d_model)）を時刻ごとにスタックし、  
     ```
     stacked = torch.stack([x_ball, x_court, x_player], dim=2)  # (B, T, 3, d)
     ```  
   - これを Flatten して (B*T, 3, d) とし、学習可能なクエリ `cross_query` を使った `MultiheadAttention` で「そのフレームにおいて 3 モダリティのどれを重視すべきか」を算出します。  
     ```python
     fused, _ = self.cross_attn(
         query=cross_query.expand(B*T, 1, d),
         key=flat_stacked,
         value=flat_stacked
     )  # fused: (B*T, 1, d)
     fused = fused.view(B, T, d)
     ```  
   - これにより「単に concat → 線形投影する」よりも、ダイナミックにモダリティ間の重みづけを学習できます。

4. **最終的に分類ヘッド**  
   - 上で融合した `(B, T, d_model)` に対して `self.classifier` を当て、`logits` を得ます（形状 `(B, T, 2)`）。

### 改修後のサンプルコード（概要）
```python
class EventTransformerV2(nn.Module):
    def __init__(self, d_model=128, nhead=8, num_layers=4, dropout=0.1, max_seq_len=512, pose_dim=51):
        super().__init__()
        # 1) 各モダリティ埋め込み
        self.ball_embed   = nn.Linear(3, d_model)
        self.court_embed  = nn.Linear(45, d_model)
        self.player_embed = nn.Linear(5 + pose_dim, d_model)

        # 2) プレイヤー Attention Pooling
        self.player_attn_q = nn.Parameter(torch.zeros(1, 1, d_model))
        self.player_attn   = nn.MultiheadAttention(embed_dim=d_model, num_heads=nhead, batch_first=True)

        # 3) 各モダリティ専用 TransformerEncoder
        enc_ball   = nn.TransformerEncoderLayer(d_model, nhead, 4*d_model, dropout, batch_first=True, norm_first=True)
        enc_court  = nn.TransformerEncoderLayer(d_model, nhead, 4*d_model, dropout, batch_first=True, norm_first=True)
        enc_player = nn.TransformerEncoderLayer(d_model, nhead, 4*d_model, dropout, batch_first=True, norm_first=True)
        self.encoder_ball   = nn.TransformerEncoder(enc_ball,   num_layers)
        self.encoder_court  = nn.TransformerEncoder(enc_court,  num_layers)
        self.encoder_player = nn.TransformerEncoder(enc_player, num_layers)

        # 4) 位置エンコーディング
        self.pos_enc = LearnablePositionalEncoding(max_seq_len, d_model)

        # 5) Cross‐Attention 融合
        self.cross_query = nn.Parameter(torch.zeros(1, 1, d_model))
        self.cross_attn  = nn.MultiheadAttention(embed_dim=d_model, num_heads=nhead, batch_first=True)

        # 6) 分類ヘッド
        self.classifier = nn.Linear(d_model, 2)

    def forward(self, ball, player_bbox, player_pose, court):
        B, T, P, _ = player_bbox.shape

        # (1) プレイヤー情報の埋め込み → Attention Pooling
        players = torch.cat([player_bbox, player_pose], dim=-1)
        emb_players = self.player_embed(players)  # (B,T,P,d)
        mask = (emb_players.abs().sum(-1) > 0)     # (B,T,P)

        flat_pl = emb_players.view(B*T, P, -1)
        flat_mk = mask.view(B*T, P)
        q   = self.player_attn_q.expand(B*T, -1, -1)  # (B*T,1,d)
        out, _ = self.player_attn(q, flat_pl, flat_pl, key_padding_mask=~flat_mk)
        player_repr = out.view(B, T, -1)               # (B,T,d)

        # (2) ball/court 埋め込み + PosEnc + 各 Encoder
        ball_repr  = self.pos_enc(self.ball_embed(ball))
        court_repr = self.pos_enc(self.court_embed(court))
        player_repr = self.pos_enc(player_repr)

        x_ball   = self.encoder_ball(ball_repr)     # (B,T,d)
        x_court  = self.encoder_court(court_repr)   # (B,T,d)
        x_player = self.encoder_player(player_repr) # (B,T,d)

        # (3) Cross‐Attention 融合
        stacked = torch.stack([x_ball, x_court, x_player], dim=2)  # (B,T,3,d)
        flat_st = stacked.view(B*T, 3, -1)                         # (B*T,3,d)
        q2 = self.cross_query.expand(B*T, -1, -1)                 # (B*T,1,d)
        fused, _ = self.cross_attn(q2, flat_st, flat_st)           # (B*T,1,d)
        fused = fused.view(B, T, -1)                               # (B,T,d)

        # (4) 分類ヘッド
        logits = self.classifier(fused)  # (B,T,2)
        return logits
```

### 期待する効果
- プレイヤー情報のマスク平均に伴う潜在的な埋もれを回避し、最も影響力のあるプレイヤー（あるいはキーポイント）を Attention Pooling で抽出。

- Ball/Court/Player を別個に自己注意処理できるため、モダリティ固有の時系列特徴を最初に獲得しやすくなる。

- Cross‐Attention で必要なモダリティだけを動的に強調できるため、例えば「ボールの軌道変化が重要なフレーム」では ball 特徴に重みを乗せ、「プレイヤーの動きが重要なフレーム」では player 特徴を強調するなど、場面に応じたモダリティ融合が可能になる。

- 結果として、「各フレームごとの hit/bounce 判定精度」が向上し、情報損失を最小限に抑えたままトランスフォーマ全体で学習できるようになります。
